In [8]:
%load_ext autoreload
%autoreload 2
import pandas as pd
import numpy as np
from ipumspy import readers, ddi
from Credentials import MyCredentials
from Functions import *
from pathlib import Path

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


# Cleaning IPUMS
I will first clean the IPUMS extracts.

## NOTE
You should really only have to run this section once. It is time consuming so I save the resulting data in "LaborByCounty.csv". If you already ran this part of the code and have this file in your data directory then skip to the next cleaning section.

In [9]:
DataByCounty = pd.DataFrame()
DataDir = Path(Paths['data'])
Files = list(DataDir.glob("*.xml"))
AggGroups = [['foreign01','fulltime01','fipcode'],
             ['foreign01','fulltime02','fipcode'],
             ['foreign02','fulltime01','fipcode'],
             ['foreign02','fulltime02','fipcode']]

for f in Files:

    # Get the Data dictionary
    ddi = readers.read_ipums_ddi(f)
    
    # Read data into data frame
    ipums_df = readers.read_microdata(ddi, DataDir / ddi.file_description.filename)
    
    print('\n**********************************************************************\n' +
    f'Working file {f} corresponding to sample ' + str(ipums_df['YEAR'][0]) +
    '\n**********************************************************************\n')

    try :

        # Clean the data - IpumsTidy() is defined in Functions.py
        ipums_tidy = IpumsTidy(ipums_df)

    except :
        print(
            '\n**********************************************************************\n' +
            f'Skipping ' + str(ipums_df['YEAR'][0]) + ' for now'
            '\n**********************************************************************\n'
        )

        continue

    # We will use this in a moment
    year = ipums_tidy['year'].iloc[0]

    for g in AggGroups:
        Aggregated = (
            WeightedSum(ipums_tidy,g) # Defined in the Functions.py file
            .query('fulltime == 1')   # Keep only full time workers
            .assign(year = year)      # Record the sample year
        )
        DataByCounty = pd.concat([DataByCounty,Aggregated])
    


/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00046.xml corresponding to sample 2019
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00047.xml corresponding to sample 2020
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00048.xml corresponding to sample 2021
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00050.xml corresponding to sample 2023
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00031.xml corresponding to sample 1950
**********************************************************************


**********************************************************************
Skipping 1950 for now
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00032.xml corresponding to sample 2005
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00033.xml corresponding to sample 2006
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00034.xml corresponding to sample 2007
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00035.xml corresponding to sample 2008
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00036.xml corresponding to sample 2009
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00037.xml corresponding to sample 2010
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00038.xml corresponding to sample 2011
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00039.xml corresponding to sample 2012
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00040.xml corresponding to sample 2013
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00041.xml corresponding to sample 2014
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00042.xml corresponding to sample 2015
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00043.xml corresponding to sample 2016
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00044.xml corresponding to sample 2017
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00045.xml corresponding to sample 2018
**********************************************************************



/home/r/remuszka/.conda/envs/ImmigrationTfp/lib/python3.12/site-packages/ipumspy/readers.py:70: CitationWarning: Use of data from IPUMS is subject to conditions including that users should cite the data appropriately.
See the `ipums_conditions` attribute of this codebook for terms of use.
See the `ipums_citation` attribute of this codebook for the appropriate citation.
  warnings.warn(



**********************************************************************
Working file ../data/usa_00049.xml corresponding to sample 2022
**********************************************************************



In [15]:
# Drop missing counties
DataByCounty = DataByCounty.pipe(lambda x: x.loc[x['fipcode'].str[-3:] != '000'])

foreign  fulltime fipcode  HoursSupplied  BodiesSupplied met2013  \
479        0         1   01003      3271716.0         73567.0   19300   
480        0         1   01015      1691339.0         36853.0   11500   
481        0         1   01055      1461909.0         33207.0   23460   
482        0         1   01073     10216184.0        227003.0   13820   
483        0         1   01081      2557082.0         57355.0   12220   

    ForeignVar HoursVar       year  
479         01       01 2019-01-01  
480         01       01 2019-01-01  
481         01       01 2019-01-01  
482         01       01 2019-01-01  
483         01       01 2019-01-01

In [10]:
DataByCounty.to_csv(Paths['data'] + '/LaborByCounty.csv', index = False)
DataByCounty.head()

foreign  fulltime fipcode  HoursSupplied  BodiesSupplied met2013  \
478        0         1   01000     44057754.0        996624.0   33860   
479        0         1   01003      3271716.0         73567.0   19300   
480        0         1   01015      1691339.0         36853.0   11500   
481        0         1   01055      1461909.0         33207.0   23460   
482        0         1   01073     10216184.0        227003.0   13820   

    ForeignVar HoursVar       year  
478         01       01 2019-01-01  
479         01       01 2019-01-01  
480         01       01 2019-01-01  
481         01       01 2019-01-01  
482         01       01 2019-01-01

# Cleaning BEA
Now I clean the BEA county-level GDP data. While we have this series from 2001 onward, the sample period is 2005 onward because the smallest level of geography avaiable in the ACS Ipums extracts is larger than a county before 2005. You can read about the samples [here](https://usa.ipums.org/usa-action/samples)

In [35]:
Bea_Tidy = (BeaTidy(pd.read_csv(Paths['data'] + '/GdpByCounty.csv', dtype={'fipcode':'object'}))
            .assign(year = lambda x: pd.to_datetime(x['year']))
)
Bea_Tidy = Bea_Tidy[Bea_Tidy['year'] >= '2005-01-01']                # Keep 2005 on
Bea_Tidy.head()

County Name     YNom       year fipcode
4  Autauga, AL  1015320 2005-01-01   01001
5  Autauga, AL  1124089 2006-01-01   01001
6  Autauga, AL  1182234 2007-01-01   01001
7  Autauga, AL  1097703 2008-01-01   01001
8  Autauga, AL  1175769 2009-01-01   01001

# Merging County Files

In [68]:
DataByCounty = (pd.read_csv(Paths['data'] + '/LaborByCounty.csv', dtype={'fipcode':'object','met2013':'object'})
                .assign(year = lambda x: pd.to_datetime(x['year']))
)
Merged = (pd.merge(Bea_Tidy,DataByCounty, on = ['fipcode','year'], how = 'left', indicator = True)
         .assign(met2013 = lambda x: x['met2013'].astype(str)) # Don't use NaN for missing values
         .pipe(lambda x: x.loc[x['_merge'] != 'left_only'])           # Keep only the matches
         .drop(columns = ['_merge'])
         )         
Merged.head()

County Name     YNom       year fipcode  foreign  fulltime  HoursSupplied  \
19  Baldwin, AL  4717641 2005-01-01   01003      0.0       1.0      2775692.0   
20  Baldwin, AL  4717641 2005-01-01   01003      1.0       1.0        97048.0   
21  Baldwin, AL  4717641 2005-01-01   01003      0.0       1.0      2910383.0   
22  Baldwin, AL  4717641 2005-01-01   01003      1.0       1.0        97048.0   
23  Baldwin, AL  4717641 2005-01-01   01003      0.0       1.0      2778332.0   

    BodiesSupplied met2013  ForeignVar  HoursVar  
19         61994.0   19300         1.0       1.0  
20          2245.0   19300         1.0       1.0  
21         65770.0   19300         1.0       2.0  
22          2245.0   19300         1.0       2.0  
23         62060.0   19300         2.0       1.0

In [69]:
print(  # How many counties are represented?
'COUNTIES ACCOUNTED FOR: ' + str(len(np.unique(Merged['fipcode']))) + '\n'
'MSAs ACCOUNTED FOR: ' + str(len(np.unique(Merged['met2013'])))
)

COUNTIES ACCOUNTED FOR: 452
MSAs ACCOUNTED FOR: 242


There are 242 MSAs in our estimation sample. Unfortunately, there really aren't that many counties...

# Merge CBSA and MSA Titles From Census

In [70]:
CbsaTitles = (pd.read_excel(Paths['data'] + '/CBSA_2013_Census_Bureau_Delineation_File.xls', header=2,
                           usecols=['CBSA Code', 'CBSA Title', 'State Name','CSA Title'],
                           dtype={'CBSA Code':'object', 'CBSA Title':'object', 'State Name':'object', 'CSA Title':'object'})
             .rename(columns={'CBSA Code':'met2013'})
             )

In [71]:
CbsaTitles.head()

met2013    CBSA Title CSA Title    State Name
0   10100  Aberdeen, SD       NaN  South Dakota
1   10100  Aberdeen, SD       NaN  South Dakota
2   10140  Aberdeen, WA       NaN    Washington
3   10180   Abilene, TX       NaN         Texas
4   10180   Abilene, TX       NaN         Texas

In [72]:
CountyData = pd.merge(Merged, CbsaTitles, on='met2013', how = 'left', indicator=True)
print( # All our MSAs were matched!
    'IN COUNTY DATA ONLY: ' + str((CountyData['_merge'] == 'left_only').sum())
)
CountyData = CountyData.drop(columns=['_merge'])

IN COUNTY DATA ONLY: 0


# Merge Price Deflator
If you haven't already, you should run the code form the cell title "Merge County Files"

In [73]:
PricesDf = (pd.read_csv(Paths['data'] + '/GdpPriceDeflator.csv')
            .assign(year = lambda x: pd.to_datetime(x['Unnamed: 0']))
            .drop(columns = ['Unnamed: 0'])
            .rename(columns = {'0':'P'})
            )
CountyData = pd.merge(CountyData,PricesDf,on='year', how = 'left', indicator=False) # All observations in CountyData Matched

# SAVING to CSV
CountyData.to_csv(Paths['data'] + '/CountyData.csv', index = False)

# Saving to .dta
CountyData = CountyData.rename(columns={'County Name':'CountyName','CBSA Title':'CbsaName', 'State Name':'StateName','CSA Title':'CsaTitle'})
CountyData.to_stata(Paths['data'] + '/CountyData.dta', write_index=False, convert_dates={'year':'ty'})

# Displaying
CountyData.head()

CountyName     YNom       year fipcode  foreign  fulltime  HoursSupplied  \
0  Baldwin, AL  4717641 2005-01-01   01003      0.0       1.0      2775692.0   
1  Baldwin, AL  4717641 2005-01-01   01003      1.0       1.0        97048.0   
2  Baldwin, AL  4717641 2005-01-01   01003      0.0       1.0      2910383.0   
3  Baldwin, AL  4717641 2005-01-01   01003      1.0       1.0        97048.0   
4  Baldwin, AL  4717641 2005-01-01   01003      0.0       1.0      2778332.0   

   BodiesSupplied met2013  ForeignVar  HoursVar                   CbsaName  \
0         61994.0   19300         1.0       1.0  Daphne-Fairhope-Foley, AL   
1          2245.0   19300         1.0       1.0  Daphne-Fairhope-Foley, AL   
2         65770.0   19300         1.0       2.0  Daphne-Fairhope-Foley, AL   
3          2245.0   19300         1.0       2.0  Daphne-Fairhope-Foley, AL   
4         62060.0   19300         2.0       1.0  Daphne-Fairhope-Foley, AL   

                     CsaTitle StateName       P  
0  Mobile-Daphne-Fairhope, AL   Alabama  81.556  
1  Mobile-Daphne-Fairhope, AL   Alabama  81.556  
2  Mobile-Daphne-Fairhope, AL   Alabama  81.556  
3  Mobile-Daphne-Fairhope, AL   Alabama  81.556  
4  Mobile-Daphne-Fairhope, AL   Alabama  81.556